Add support for Dnx Core [WIP]

jeremymeng opened the pull request:
This is for #202. We are still working on a few remaining items.  But I feel it's better to send this out and get feedback earlier.

Open issues
- [ ] CI builds failed because NUnit 3 is not supported in AppVeyor yet. Need workaround.
    - ~~The commands at
      https://gist.github.com/jeremymeng/06e881f9981acbd8b504 seem
      enough to have `nunit-console.exe` in `%PATH%`.~~  NUnit 3
      should work now (https://github.com/appveyor/ci/issues/498).
- [ ] Performance issues
    - [ ] Currently `BufferedStream` isn't available in .Net Core.  This caused > 10x slowdown. it is being added back (tracked by https://github.com/dotnet/corefx/issues/1793).  Once that's available (`1.0.0-rc2`) we can re-enable `BufferedStream` for `DNXCORE50`.
    - [ ] Remaining ~2x slowdown is under investigation.
    - [ ] When running `MassiveBulkOpsSync(bool preserveOrder, int threads)`, increasing number of threads doesn't help much on the .Net Core port, not like the original code.
- [ ] `Stream.BeginRead()` and `Stream.EndRead()` are not supported in .Net Core.  Which of the following options is better?
    1. I've added extension methods to simulate the pattern.  I didn't want to make big changes so I went with extension methods.    
    2. Another option is to rewrite the code using the async api Stream.ReadAsnc(). @deepakverma had an implementation at [his previous port](https://github.com/deepakverma/StackExchange.Redis/blob/client_for_k/StackExchange.Redis_NET50/src/StackExchange.Redis_NET50/StackExchange/Redis/PhysicalConnection.cs#L587).  Is this a concern for the .Net 40 projects?
- [ ] Re-visit `Type.GetTypeCode(Type)` after moving to RC.
- [x] `PLAT_SAFE_CONTINUATIONS` doesn't seem to be available on Core. **Edit** it is available.  I was using 4.0.0 version of `System.Thread.Tasks`.  Fixed.
- [ ] Move to RC. There are some breaking changes between ASP .NET beta8 and RC.
    - https://github.com/aspnet/Announcements/issues/98
- [ ] The dnx test project is a console-based project because it's the only supported way at this point.  NUnit will be adding a dnx console runner and VS integration.
- [ ] `Moq` based tests are all passing using a private build of package that supports .Net Core, but currently disabled for dnx test project.  It can be re-enabled by defining `FEATURE_MOQ` in `project.json`.
- [x] `BookSleeve` related tests were disabled.
    - It looks like there are already similar tests in the StackExchange.Redis.Tests project.
- [ ] We were not able to run the following tests as they require special setup.
    - SSDB
    - SSL
    - VPNTest
    
@mgravell 
CC @rustd @deepakverma @rogerbestmsft

mgravell commented:
Wow, that's huge. And very much appreciated. I'm going to have to dedicate
some time to reading through this lot. Great to hear that BufferedStream
should be back. I need to review the changes to comment on the rest.

Kudos
On 16 Nov 2015 9:01 p.m., "Jeremy Meng" <notifications@github.com> wrote:

> This is for #202
> <https://github.com/StackExchange/StackExchange.Redis/issues/202>. We are
> still working on a few remaining items. But I feel it's better to send this
> out and get feedback earlier.
>
> Open issues
>
>    - Performance issues
>       - Currently BufferedStream isn't available in .Net Core. This
>       caused > 10x slowdown. it is being added back (tracked by
>       dotnet/corefx#1793 <https://github.com/dotnet/corefx/issues/1793>).
>       Once that's available (1.0.0-rc2) we can re-enable BufferedStream
>       for DNXCORE50.
>       - Remaining ~2x slowdown is under investigation.
>    - Stream.BeginRead() and Stream.EndRead() are not supported in .Net
>    Core. Which of the following options is better?
>       1. I've added extension methods to simulate the pattern. I didn't
>       want to make big changes so I went with extension methods.
>       2. Another option is to rewrite the code using the async api
>       Stream.ReadAsnc(). @deepakverma <https://github.com/deepakverma>
>       had an implementation at his previous port
>       <https://github.com/deepakverma/StackExchange.Redis/blob/client_for_k/StackExchange.Redis_NET50/src/StackExchange.Redis_NET50/StackExchange/Redis/PhysicalConnection.cs#L587>.
>       Is this a concern for the .Net 40 projects?
>    - Move to RC. There are some breaking changes between ASP .NET beta8
>    and RC.
>       - aspnet/Announcements#98
>       <https://github.com/aspnet/Announcements/issues/98>
>
> @mgravell <https://github.com/mgravell>
> CC @rustd <https://github.com/rustd> @deepakverma
> <https://github.com/deepakverma> @rogerbestmsft
> <https://github.com/rogerbestmsft>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/StackExchange/StackExchange.Redis/pull/304
> Commit Summary
>
>    - Add StackExchange.Redist_NetCore project.
>    - Add FEATURE_SERIALIZATION constant.
>    - Replace Hashtable with Dictionary.
>    - Replace Array.ConvertAll usage with Linq Select().ToArray().
>    - Add an Asp.Net 5 project. The PCL seems unable to find many
>    dependencies.
>    - Ignore certain files from VS2015.
>    - Disable Browsable attribute which is from winforms for design time
>    property window support.
>    - Remove usage of unsupported BufferedStream for netcore.
>    - Disable logging of ThreadPool stats for netcore since
>    Get{Max,Min,Available}Threads() are not supported.
>    - Exclude platform-specific code for NetCore. Remove usage of
>    unsupported IClone interface.
>    - Use Type.GetTypeInfo().IsValueType for netcore.
>    - Use RegexOptions.Compiled if it is supported.
>    - Use Volatile.{Read,Write} instead.
>    - Replace StringComparer.InvariantCultrueIgnoreCase with
>    STringComparer.OrdinalIgnoreCase as the former is almost always incorrect
>    for string equality.
>    - Add package for Timer. Use SpinWait struct to simulate the
>    Thread.Yield. According to MSDN it will always yield after the initial
>    several spins. Simple test shows it yields after 11 spins.
>    - Add extension methods for Stream's APM pattern.
>    - Add GetTypeCode() extension methods for dnxcore50.
>    - Add the test project for dnxcore.
>    - Remove usage of SslStream.IsEncrypted as it is not availalbe on
>    dnxcore. On Desktop it simply returns IsAuthenticated.
>    - Rewrite to use GetBytes() overload available on dnxcore.
>    - Disable socket polling for dnxcore.
>    - Fix Volatile-related build errors for other non-dnxcore projects.
>    - Fix a runtime error. Also bump the version to 1.1.0.
>    - Updated to NUnit 3.0.0 Beta 5
>    - Merge branch 'UpdateNUnitTo3' of
>    https://github.com/RogerBestMsft/StackExchange.Redis into netcore-test
>    - Add dependencies.
>    - Replace Array.ConvertAll() with Linq Select() calls.
>    - Disable BookSleeve tests for dnxcore.
>    - Remove usage of Environment.MachineName and
>    StringComparison.InvariantCultrueIgnoreCase.
>    - Disable Thread.Abort() usage for dnxcore.
>    - Fix test project to build with dnxcore.
>    - Update to NUnit 3.0. The ExpectedExceptionAttribute is gone.
>    - Use nunitlite to run the tests.
>    - Use DNXCORE50 constant instead of NETCORE. Socket is only supported
>    on dnxcore.
>    - Keep the exception throwing behavior close to the desktop
>    implementation.
>    - Define PLAT_SAFE_CONTINUATIONS constant. Also fix NETCORE that were
>    missed in previous commit.
>    - Revert the HashTable => Dictionary<> change.
>    - Add 'static' because it is now required by NUnit 3.0.
>    - Simplify Assert.Throws usage.
>    - Update moq.netcore to 4.4.0. The old version didn't specify its
>    dependencies for the 'dotnet' group.
>    - Switch to the official NUnit package. Disable Moq-based tests for
>    now.
>    - Upgrade NUnit to 3.0.0-rc-2.
>    - Add the dnxcore50 version of BasicTest project.
>    - Add Debug and Release configurations.
>    - Remove "VERBOSE" that was added for testing only.
>    - Merge changes from upstream. Looks like PLAT_SAFE_CONTINUATIONS is
>    not available in beta8 yet.
>    - Replace usage of SpinWait struct with simple increment operation.
>    SpinWait.SpinOnce() takes too much time than the original
>    Thread.SpinWait(5).
>    - Add condition of DNXCORE50. Only check completed result.
>    - Add FEATURE_SERIALIZATION and FEATURE_SOCKET_MODE_POLL.
>
> File Changes
>
>    - *M* .gitignore
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-0>
>    (40)
>    - *M* BasicTest/Program.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-1>
>    (8)
>    - *A* BasicTest_dnxcore50/BasicTest_dnxcore50.xproj
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-2>
>    (19)
>    - *A* BasicTest_dnxcore50/project.json
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-3>
>    (41)
>    - *M* MigratedBookSleeveTestSuite/Config.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-4>
>    (39)
>    - *M* MigratedBookSleeveTestSuite/Issues/SO10504853.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-5>
>    (37)
>    - *M* MigratedBookSleeveTestSuite/MigratedBookSleeveTestSuite.csproj
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-6>
>    (6)
>    - *M* MigratedBookSleeveTestSuite/Program.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-7>
>    (30)
>    - *M* MigratedBookSleeveTestSuite/Scripting.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-8>
>    (25)
>    - *M* MigratedBookSleeveTestSuite/packages.config
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-9>
>    (2)
>    - *M* StackExchange.Redis.Tests/BasicOps.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-10>
>    (1395)
>    - *M* StackExchange.Redis.Tests/BatchWrapperTests.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-11>
>    (7)
>    - *M* StackExchange.Redis.Tests/Cluster.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-12>
>    (1126)
>    - *M* StackExchange.Redis.Tests/Config.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-13>
>    (13)
>    - *M* StackExchange.Redis.Tests/DatabaseWrapperTests.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-14>
>    (10)
>    - *M* StackExchange.Redis.Tests/Issues/Issue25.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-15>
>    (14)
>    - *M* StackExchange.Redis.Tests/Issues/SO25567566.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-16>
>    (2)
>    - *M* StackExchange.Redis.Tests/Locking.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-17>
>    (2)
>    - *M* StackExchange.Redis.Tests/MultiMaster.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-18>
>    (19)
>    - *M* StackExchange.Redis.Tests/Naming.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-19>
>    (61)
>    - *M* StackExchange.Redis.Tests/Profiling.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-20>
>    (7)
>    - *M* StackExchange.Redis.Tests/PubSub.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-21>
>    (48)
>    - *M* StackExchange.Redis.Tests/Scripting.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-22>
>    (366)
>    - *M* StackExchange.Redis.Tests/Secure.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-23>
>    (14)
>    - *M* StackExchange.Redis.Tests/Sentinel.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-24>
>    (4)
>    - *M* StackExchange.Redis.Tests/StackExchange.Redis.Tests.csproj
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-25>
>    (16)
>    - *M* StackExchange.Redis.Tests/TestBase.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-26>
>    (9)
>    - *M* StackExchange.Redis.Tests/TransactionWrapperTests.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-27>
>    (7)
>    - *M* StackExchange.Redis.Tests/WithKeyPrefixTests.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-28>
>    (256)
>    - *M* StackExchange.Redis.Tests/WrapperBaseTests.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-29>
>    (12)
>    - *M* StackExchange.Redis.Tests/packages.config
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-30>
>    (2)
>    - *A*
>    StackExchange.Redis.Tests_dnxcore50/StackExchange.Redis.Tests/Program.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-31>
>    (14)
>    - *A*
>    StackExchange.Redis.Tests_dnxcore50/StackExchange.Redis.Tests/StackExchange.Redis.Tests.xproj
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-32>
>    (19)
>    - *A*
>    StackExchange.Redis.Tests_dnxcore50/StackExchange.Redis.Tests/project.json
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-33>
>    (47)
>    - *A* StackExchange.Redis.dnx.sln
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-34>
>    (34)
>    - *M* StackExchange.Redis/StackExchange.Redis.StrongName.csproj
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-35>
>    (402)
>    - *M* StackExchange.Redis/StackExchange.Redis.csproj
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-36>
>    (12)
>    - *M* StackExchange.Redis/StackExchange/Redis/CommandMap.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-37>
>    (445)
>    - *A* StackExchange.Redis/StackExchange/Redis/Compat/VolatileWrapper.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-38>
>    (29)
>    - *M* StackExchange.Redis/StackExchange/Redis/CompletionManager.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-39>
>    (17)
>    - *M* StackExchange.Redis/StackExchange/Redis/ConfigurationOptions.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-40>
>    (24)
>    - *M* StackExchange.Redis/StackExchange/Redis/ConnectionMultiplexer.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-41>
>    (44)
>    - *M* StackExchange.Redis/StackExchange/Redis/ExtensionMethods.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-42>
>    (253)
>    - *M* StackExchange.Redis/StackExchange/Redis/HashEntry.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-43>
>    (196)
>    - *A*
>    StackExchange.Redis/StackExchange/Redis/InternalRegexCompiledOption.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-44>
>    (36)
>    - *M* StackExchange.Redis/StackExchange/Redis/LuaScript.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-45>
>    (2)
>    - *M* StackExchange.Redis/StackExchange/Redis/Message.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-46>
>    (28)
>    - *M* StackExchange.Redis/StackExchange/Redis/PhysicalBridge.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-47>
>    (2)
>    - *M* StackExchange.Redis/StackExchange/Redis/PhysicalConnection.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-48>
>    (107)
>    - *M* StackExchange.Redis/StackExchange/Redis/RedisResult.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-49>
>    (781)
>    - *M* StackExchange.Redis/StackExchange/Redis/RedisValue.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-50>
>    (61)
>    - *M* StackExchange.Redis/StackExchange/Redis/ResultProcessor.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-51>
>    (2)
>    - *M* StackExchange.Redis/StackExchange/Redis/ScriptParameterMapper.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-52>
>    (14)
>    - *M* StackExchange.Redis/StackExchange/Redis/ServerEndPoint.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-53>
>    (1390)
>    - *M* StackExchange.Redis/StackExchange/Redis/SocketManager.NoPoll.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-54>
>    (2)
>    - *M* StackExchange.Redis/StackExchange/Redis/SocketManager.Poll.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-55>
>    (8)
>    - *M* StackExchange.Redis/StackExchange/Redis/SocketManager.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-56>
>    (8)
>    - *M* StackExchange.Redis/StackExchange/Redis/SortedSetEntry.cs
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-57>
>    (250)
>    - *M*
>    StackExchange.Redis_Net40/StackExchange.Redis_Net40.StrongName.csproj
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-58>
>    (400)
>    - *M* StackExchange.Redis_Net40/StackExchange.Redis_Net40.csproj
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-59>
>    (394)
>    - *M*
>    StackExchange.Redis_Net46/StackExchange.Redis_Net46.StrongName.csproj
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-60>
>    (10)
>    - *M* StackExchange.Redis_Net46/StackExchange.Redis_Net46.csproj
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-61>
>    (10)
>    - *A*
>    StackExchange.Redis_dnxcore50/StackExchange.Redis/StackExchange.Redis.xproj
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-62>
>    (19)
>    - *A* StackExchange.Redis_dnxcore50/StackExchange.Redis/project.json
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-63>
>    (60)
>    - *A* global.json
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-64>
>    (8)
>    - *D* packages/NUnit.2.6.4/lib/nunit.framework.dll
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-65>
>    (0)
>    - *D* packages/NUnit.2.6.4/lib/nunit.framework.xml
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-66>
>    (0)
>    - *D* packages/NUnit.2.6.4/license.txt
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-67>
>    (15)
>    - *A* packages/NUnit.3.0.0-rc-2/CHANGES.txt
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-68>
>    (0)
>    - *A* packages/NUnit.3.0.0-rc-2/LICENSE.txt
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-69>
>    (0)
>    - *A* packages/NUnit.3.0.0-rc-2/NOTICES.txt
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-70>
>    (0)
>    - *A* packages/NUnit.3.0.0-rc-2/lib/dotnet/nunit.framework.xml
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-71>
>    (0)
>    - *A* packages/NUnit.3.0.0-rc-2/lib/net20/nunit.framework.xml
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-72>
>    (0)
>    - *A* packages/NUnit.3.0.0-rc-2/lib/net40/nunit.framework.xml
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-73>
>    (0)
>    - *A* packages/NUnit.3.0.0-rc-2/lib/net45/nunit.framework.xml
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-74>
>    (0)
>    - *A*
>    packages/NUnit.3.0.0-rc-2/lib/portable-net45+win8+wp8+wpa81+Xamarin.Mac+MonoAndroid10+MonoTouch10+Xamarin.iOS10/nunit.framework.xml
>    <https://github.com/StackExchange/StackExchange.Redis/pull/304/files#diff-75>
>    (0)
>
> Patch Links:
>
>    - https://github.com/StackExchange/StackExchange.Redis/pull/304.patch
>    - https://github.com/StackExchange/StackExchange.Redis/pull/304.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/StackExchange/StackExchange.Redis/pull/304>.
>

stebet commented:
Wow. This is awesome! If there are any issues that are up for grabs, please tag them as such. I'd love to help :)

ForkBug commented:
Thank you for the hard working!

mgravell commented on StackExchange.Redis/StackExchange.Redis.csproj:
```diff
@@ -83,6 +84,7 @@
     <Compile Include="StackExchange\Redis\HashEntry.cs" />
     <Compile Include="StackExchange\Redis\IConnectionMultiplexer.cs" />
     <Compile Include="StackExchange\Redis\InternalErrorEventArgs.cs" />
+    <Compile Include="StackExchange\Redis\InternalRegexCompiledOption.cs" />
```
This is actually a really cute solution to the compiled regex problem; I had a similar problem in a few other libs, and just settled for 5.2 vs 5.4 + dnx451 builds, but I'm very tempted to steal this, if you don't mind. Kudos.

jeremymeng replied:
Credits to @sharwell https://github.com/dotnet/corefx/issues/340#issuecomment-120749951

mgravell commented:
Review complete; a couple of big ones I'd like to hear more about, but the biggest is the introduction of TPL into the socket-read - I really really want to understand what that is about, because on the surface that seems incredibly bad.

Much more minor thoughts, summarized (I may have missed one):

- the odd change of case sensitivity
- the ConvertAll would be preferred to right-size allocate (sample code provided)
- would be really nice if this was rc1 compatible rather than beta-8
- the tests don't seem to run for me at all
- the allocations around BeginRead - which is tied into the major socket-read issue

jeremymeng commented:
> the odd change of case sensitivity

Fixed

> the ConvertAll would be preferred to right-size allocate (sample code provided)

Fixed

> would be really nice if this was rc1 compatible rather than beta-8

I have been working on it.  Other than bumping versions, the only issue is that in RC1 `Socket.BeginConnect()` is gone as well.  So the situation is similar to `BeginRead()/EndRead()`:  I am not confident enough to rewrite the code using `ConnectAsync()` while keeping the same behavior (exception, logging, etc.)

Maybe you could rewrite `master` branch to use Async Api's instead? Then this port would get them for free.

> the tests don't seem to run for me at all

Were you able to run the project, but got a crash?  I think it might a NUnit bug where some exception are unhandled.   Due to the performance issue, the `BasicOpsTests` could throw `TimeoutException` which will crash dnx. I was able to run that test after modifying timeout limits.

Another test issue is that some `AggregateException` from the Async Api were not handled thus show up in the ambient exceptions during `TearDown()`.

> the text encoding chunk

I noticed that and even attempted to fix it but it didn't do anything good.  Everything looks normal in VS.  Maybe I should use another editor, or adjust my git `crlf` option?  I have it set to `core.autocrlf=true` now.  Will give it another try.


> PLAT_SAFE_CONTINUATIONS just triggers TaskCreationOptions.RunContinuationsAsynchronously, which is declared - does it not work?

Fixed.

> BufferedStream - I don't mind writing a version of that; shouldn't be too tricky especially as it is write only

It's coming soon...

> I would wager $5 that the async-over-sync-over-async is a good part of the rest of the slowdown (that and the extra allocations); any takers?

I have seen the Task overhead in perfview traces.  I have not captured a trace with the `Stream.ReadAsync()` version. I think without the `BufferedStream` fix this overhead is just masked by the regression so it might make sense to investigate later.

mgravell commented:
I have merged this into the branch on SE.Redis: https://github.com/StackExchange/StackExchange.Redis/tree/jeremymeng-netcore - and updated to the latest bits (and made the project files more manageable, etc). There is a problem with the Socket.Connect API - looks like this changed between beta8 and rc1; I will try to take a look

mgravell commented:
(happy to do an alpha in the current state, even though there are outstanding problems, but we need to get that Connect API working)

mgravell commented:
Note: I've also changed project.json to target net40, net45, net46 and dotnet5.5; I would *like* to try to target down-level netstandard, but there are a few problems there, too; 5.4 fails because of `Socket.Available`, for example

mgravell commented:
I've added something that *probably* fixes the connect problem. There is still something **very** wrong about the `BeginRead` code; that *really* needs to be refactored - we cannot keep (past alpha) something that does async over sync over async. Good enough for alpha, but that's all.

rustd commented:
Thank you Marc. How do you want to track the remaining work?

mgravell commented:
I think I've fixed the StreamExtensions (BeginRead etc) code - would appreciate some eyeballs. I've also enabled direct dnx451 comparison in the basic test, and the terrible news is: it is 50-100 times slower. I'm hoping a lot of this is `BufferedStream` !

The current code seems to at least pass the most basic of operations. I'm happier to issue an alpha with the BeginRead stuff changed, which I will do now; however, I have **not** done extensive testing (I'm up to my neck in 20 different things, plus home life is ... complicated and busy).

My biggest concern right now is the performance. That is going to take some digging.

mgravell commented:
Issued:

https://www.nuget.org/packages/StackExchange.Redis/1.1.0-alpha1
https://www.nuget.org/packages/StackExchange.Redis.StrongName/1.1.0-alpha1

Now that we have something there, we should be able to do a **much** faster turnaround if we need to do an alpha2-alpha347, etc

jeremymeng commented:
Thanks a lot Marc!  Should we close this PR and do additional work against the other branch?

mgravell commented:
I was hoping it would do that when I pulled it into a branch, but I guess the tools only do that for master. I've changed the name of the branch to `CoreCLR` for convenience, but it is all there. I need to go through the PR again and create a new issue with a checklist of things we need to look at.

Thanks for all your work making this happen. There *is* still quite a bit of work needed, but it is great to have a release - you've saved me a lot of time and frustration.

mgravell commented:
Note: I've introduced `BufferedOutputStream` for the core-clr build, which fixes the performance problem, but there seems to be an intermittent glitch that causes problems connecting.

Performance on dnx451:

After 10000: 10000
(with continuation, preserve order)
MassiveBulkOpsAsync: Time for 10000 ops: 27ms; ops/s: 367595.703541417
After 10000: 10000
(no continuation, preserve order)
MassiveBulkOpsAsync: Time for 10000 ops: 10ms; ops/s: 930249.86511377
After 10000: 10000
(with continuation, any order)
MassiveBulkOpsAsync: Time for 10000 ops: 20ms; ops/s: 482516.03159515
After 10000: 10000
(no continuation, any order)
MassiveBulkOpsAsync: Time for 10000 ops: 10ms; ops/s: 942649.221371743

Performance on dnxcore50

After 10000: 10000
(with continuation, preserve order)
MassiveBulkOpsAsync: Time for 10000 ops: 46ms; ops/s: 214035.140289333
After 10000: 10000
(no continuation, preserve order)
MassiveBulkOpsAsync: Time for 10000 ops: 10ms; ops/s: 934169.103291078
After 10000: 10000
(with continuation, any order)
MassiveBulkOpsAsync: Time for 10000 ops: 17ms; ops/s: 563783.664932092
After 10000: 10000
(no continuation, any order)
MassiveBulkOpsAsync: Time for 10000 ops: 31ms; ops/s: 314012.164831266

This is such a massive performance difference that I'm going to issue alpha2

jeremymeng commented:
Wonderful news!  Sounds like the perf regression is mostly due to lacking `BufferedStream`, and task overhead in my implementation.
I just submit a PR for test change to ignore exceptions that doesn't seem to be caught and ignored by the async implementation.  Could you please take a look?

mgravell commented:
Am at a family weekend, and have a PDA course on Monday. Will look Tuesday.
On 11 Dec 2015 8:18 pm, "Jeremy Meng" <notifications@github.com> wrote:

> Wonderful news! Sounds like the perf regression is mostly due to lacking
> BufferedStream, and task overhead in my implementation.
> I just submit a PR for test change to ignore exceptions that doesn't seem
> to be caught and ignored by the async implementation. Could you please take
> a look?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/StackExchange/StackExchange.Redis/pull/304#issuecomment-164038538>
> .
>

noahvans commented:
Please note that dnx doesn't support Sockets.IOControl() method for Linux\Unix now.

code is here:
https://github.com/dotnet/corefx/blob/master/src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs#L970

Since StackExchange.Redis.SockectManager.SetFastLoopbackOption method is still using it now, so we cannot make these code run on Linux/Unix Platform to connect to any redis server.

mgravell commented:
K, I'll see if we can try/catch around that. The full .net version actually
does OS sniffing on that line, but the OS sniffing it uses: *also* isn't
available in Core CLR. Such fun!
On 21 Dec 2015 10:50 a.m., "Noah Vans" <notifications@github.com> wrote:

> Please note that dnx doesn't support Sockets.IOControl() method for
> Linux\Unix now.
>
> code is here:
>
> https://github.com/dotnet/corefx/blob/master/src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs#L970
>
> Since StackExchange.Redis.SockectManager.SetFastLoopbackOption method is
> still using it now, so we cannot make these code run on Linux/Unix Platform
> to connect to any redis server.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/StackExchange/StackExchange.Redis/pull/304#issuecomment-166269123>
> .
>

jeremymeng commented:
@mgravell @noahvans 
I submitted a PR to add the try-catch around socket.IOControl() call.  Please review when you get a chance.  Thanks!

jeremymeng commented:
Since the commits have been merged into CoreClr branch.  I am closing this PR.  Open work items are copied to https://github.com/StackExchange/StackExchange.Redis/issues/328.
